### PR TITLE
Dev-6446 Missing Account Balance Data

### PR DESCRIPTION
**Pull request by jonhill13**
**High level description:**

Adds Missing Account Balance Data Modal

**Technical details:**

> • Add Missing Account Balance Container

> • New function in about the data helper to format table data

> • Updates css for table specific styling

**JIRA Ticket:**
[DEV-6446](https://federal-spending-transparency.atlassian.net/browse/DEV-6446)

**Mockup:**
https://bahdigital.invisionapp.com/share/T9IAFM1NCJU#/296052496_Modal__TAS

The following are ALL required for the PR to be merged:

Author:
- [x] Linked to this PR in JIRA ticket
- [x] Scheduled demo including Design/Testing/Front-end OR Provided instructions for testing in JIRA and PR `if applicable`
- [x] Verified cross-browser compatibility: Chrome, Safari, Firefox, Internet Explorer, Edge
- [x] Verified mobile/tablet/desktop/monitor responsiveness
- [x] Verified that this PR does not create any *new* accessibility issues (via Axe Chrome extension)
- [x] Added Unit Tests for helper functions, reducers, models and Container/Component Interactivity Expectations `if applicable` [React Testing Library](react-testing-library.md)
- [N/A] [API contract](https://github.com/fedspendingtransparency/usaspending-api/tree/dev/usaspending_api/api_contracts) updated `if applicable`
- [N/A] [Component Library Integration Status Report](https://github.com/fedspendingtransparency/data-act-documentation/blob/data-transparency-ui/frontend_apps/component-library-integration-status.md) updated `if applicable`

Reviewer(s):
- Note: this PR includes this PR https://github.com/fedspendingtransparency/usaspending-website/pull/2236 and has all the styling changes from design review.
- [x] Code review complete

### CODE REVIEW

File: src/js/containers/aboutTheData/modals/MissingAccountBalanceContainer.jsx [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+
+    // on mount fetch data, and unmount cleanup pubDatesRequest
+    useEffect(() => {
+        missingAccountBalancesRequest();
+        return () => {
+            if (missingAccBalancesRequest.current) missingAccBalancesRequest.current.cancel();
+        };
+    }, []);
+    // on page, sort, order, or limit change fetch new data
+    useEffect(() => {
```
Comment (ebdabbs) [inline src/js/containers/aboutTheData/modals/MissingAccountBalanceContainer.jsx:79]: To be consistent with the behavior of other tables, it should reset to page 1 when the sort or limit changes ([example here](https://github.com/fedspendingtransparency/usaspending-website/blob/dev/src/js/containers/covid19/recipient/SpendingByRecipientContainer.jsx#L282-L294))

File: src/js/containers/aboutTheData/modals/PublicationDatesContainer.jsx [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+    }, [submissionPeriods]);
+    // on mount fetch data, and unmount cleanup pubDatesRequest
+    useEffect(() => {
+        publicationDatesRequest();
+        return () => {
+            if (pubDatesRequest.current) pubDatesRequest.current.cancel();
+        };
+    }, []);
+    // on page, sort, order, or limit change fetch new data
+    useEffect(() => {
```
Comment (ebdabbs) [inline src/js/containers/aboutTheData/modals/PublicationDatesContainer.jsx:107]: Similar to above (reset to page 1)

File: src/js/dataMapping/aboutTheData/modals.js [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+                else {
+                    const data = {
+                        page_metadata: pageMetaData,
+                        results: mockDataPublicationDates.slice(params.limit, params.limit * 2)
+                    };
+                    resolve({ data });
+                }
+            }, 1000);
+        }),
+        cancel: () => console.log(' :wave: ')
```
Comment (ebdabbs) [inline src/js/dataMapping/aboutTheData/modals.js:187]: 👋

File: src/js/helpers/aboutTheDataHelper.js [hunk 4]
```diff
@@ -0,0 +1,60 @@
+/**
+ * aboutTheDataHelper.js
+ * Created by Jonathan Hill 11/20/20
+ */
+
+import { calculateTreemapPercentage, formatMoney } from 'helpers/moneyFormatter';
+import { apiRequest } from './apiRequest';
+
+export const aboutTheDataQueryString = (params) => {
```
Comment (ebdabbs) [inline src/js/helpers/aboutTheDataHelper.js:9]: Nice!

File: src/js/helpers/aboutTheDataHelper.js [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+    return data.results.map((tasData) => {
+        let amount = '--';
+        let percent = '--';
+        if (typeof tasData.amount === 'number' && weHaveTotalData) percent = calculateTreemapPercentage(tasData.amount, data.totalObligationsNotInGTAS);
+        if (typeof tasData.amount === 'number') amount = formatMoney(tasData.amount);
+        return [tasData.tas, amount, percent];
+    });
+};
+
+export const showQuarterText = (period) => [3, 6, 9, 12].includes(period);
```
Comment (ebdabbs) [inline src/js/helpers/aboutTheDataHelper.js:60]: ✨

File: tests/helpers/aboutTheDataHelper-test.js [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+            expect(data[0][2]).toBe('5.2%');
+            expect(data[1][1]).toBe('$0');
+            expect(data[1][2]).toBe('0.0%');
+        });
+    });
+    describe('showQuarterText', () => {
+        it('should show quarter text for quarter one', () => {
+            expect(showQuarterText(3)).toBeTruthy();
+        });
+        it('should show quarter text for quarter one', () => {
```
Comment (ebdabbs) [inline tests/helpers/aboutTheDataHelper-test.js:124]: These descriptions should be updated (`should show quarter text for quarter two` etc)

File: tests/redux/reducers/account/accountReducer-test.js [hunk 7]
```diff
@@ -191,11 +192,11 @@ describe('accountReducer', () => {
         const mockActionState = {
             objectClass: mockOCAPI.results,
             objectClassDefinitions: {
-                '1': 'First Major',
-                '2': 'First Major Minor'
+                1: 'First Major',
```
Comment (ebdabbs) [inline tests/redux/reducers/account/accountReducer-test.js:195]: Did these change from strings to numbers?

**Review by ebdabbs — CHANGES_REQUESTED** [review]

File: src/js/dataMapping/aboutTheData/modals.js (hunk 3 above)
Comment (jonhill13) [reply src/js/dataMapping/aboutTheData/modals.js:187]: lol

File: tests/redux/reducers/account/accountReducer-test.js (hunk 7 above)
Comment (jonhill13) [reply tests/redux/reducers/account/accountReducer-test.js:195]: I did not make those changes, i think that came from my linter autosave making corrections.

**Review by ebdabbs — APPROVED** [review]

**Pull request merged**